framework/include/base/MooseObject.h

```diff
@@ -32,6 +82,8 @@ InputParameters validParams<MooseObject>();
  * Every object that can be built by the factory should be derived from this class.
  */
 class MooseObject :
+  public virtual BlockRestrictableHelper, // see BlockRestrictable.h
+  public virtual BoundaryRestrictableHelper, // see BoundaryRestrictable.h
```

Comment:
I knew this was going to be rejected, I don't like it much either. I will take another crack at fixing it: It was motivated by cyclic inheritance: FEProblem->NonlinearSystem->AuxKernelWarehouse->AuxKernel->BlockRestrictable->FEProblem

But, perhaps making the warehouses UniquePtr fixed this problem, I did that after I did this virtual inheritance stuff.